login/service.go

```diff
@@ -64,7 +64,12 @@ type KeycloakOAuthProvider struct {
 // KeycloakOAuthService represents keycloak OAuth service interface
 type KeycloakOAuthService interface {
 	Perform(ctx *app.LoginLoginContext, config oauth.OauthConfig, serviceConfig LoginServiceConfiguration) error
+	PerformAuthorize(ctx *app.AuthorizeAuthorizeContext, config oauth.OauthConfig, serviceConfig LoginServiceConfiguration) error
+	GetTokenFromAuthorizationCode(ctx context.Context, code string, config oauth.OauthConfig) (*oauth2.Token, error)
+	VerifyState(ctx context.Context, state string, code string) (*url.URL, error)
 	CreateOrUpdateIdentity(ctx context.Context, accessToken string, configuration LoginServiceConfiguration) (*account.Identity, bool, error)
+	CreateOrUpdateIdentityAndUser(ctx context.Context, code string, referrerURL *url.URL, keycloakToken *oauth2.Token, request *goa.RequestData, config oauth.OauthConfig, serviceConfig LoginServiceConfiguration) (*string, error)
+	BeforeRedirectToLogin(ctx context.Context, redirect *string, link *bool, apiClient *string, request *goa.RequestData, config oauth.OauthConfig, serviceConfig LoginServiceConfiguration) (*string, error)
 	Link(ctx *app.LinkLinkContext, brokerEndpoint string, clientID string, validRedirectURL string) error
 	LinkSession(ctx *app.SessionLinkContext, brokerEndpoint string, clientID string, validRedirectURL string) error
 	LinkCallback(ctx *app.CallbackLinkContext, brokerEndpoint string, clientID string) error
```

Comment:
I would suggest:
1. Rename Perform() -> Login()
2. PerformAuthorize() -> AuthCodeURL() and just return an URL here or error. The controller will create a proper redirect response. See token/link/link.go for example. Login() could re-use AuthCodeURL().
3. GetTokenFromAuthorizationCode() -> Exchange() 
So, AuthCodeURL() and Exchange() will match names used by oauth2.Config
4. VerifyState() do you need to export it? Looks like an internal service method. Create a new AuthCodeCallback() and call it from the controller. You can verify state inside this method so you don't have to expose it. Similar to what we have in token/link/link.go
5. VerifyState() is not accurate anyway. This method is actually removing the existing state too. Maybe something like reclaimReferrer() or something like this?
6. BeforeRedirectToLogin() why it's exported? Why it's in the interface? Looks like an internal method. saveAuthParams() ?
7. CreateOrUpdateIdentityAndUser vs CreateOrUpdateIdentity is misleading. Both methods create both user and identity. Maybe we should rename CreateOrUpdateIdentity() to CreateOrUpdateIdentityInDB() ? @sbose78 WDYT?

Reply:
2. We have a AuthCodeURL() method in OauthConfig, wont it be confusing?
5. reclaimReferrer() seems like a bettter name, but I dont see how it is removing existing state?
7. Probably we should rename the new method CreateOrUpdateIdentityAndUser() instead?

Reply:
@alexeykazakov sounds good - now that we aren't using the *linking during login* anymore, CreateOrUpdateIdentityInDB() looks fine. 

@kishansagathiya 
> 5 -> reclaimReferrer() seems like a bettter name, but I dont see how it is removing existing state?
the problem with the name `VerifyState()` was that it looked read-only and it does a lot more actually. 
`reclaimReferrer()` ensures the illusion of read-only is gone :)

> 2 -> We have a AuthCodeURL() method in OauthConfig, wont it be confusing?
Should be fine since the interfaces are different.
and the idea is to actually *match* the names with OauthConfig for uniformity's sake. :)

Reply:
https://github.com/fabric8-services/fabric8-auth/pull/210/commits/ec079fc047eefd3dff08b139dd5344b90e52defe